libs/MeshKernel/src/Mesh.cpp

```diff
@@ -1171,3 +1232,107 @@ bool Mesh::IsValidEdge(const UInt edgeId) const
            m_nodes[m_edges[edgeId].first].IsValid() && m_nodes[m_edges[edgeId].second].IsValid();
 }
+
+//--------------------------------
+
+void Mesh::CommitAction(const AddNodeAction& undoAction)
+{
+    m_nodes[undoAction.NodeId()] = undoAction.Node();
+    // TODO is this necessary, will it be set in Administrate?
+    m_nodesNumEdges[undoAction.NodeId()] = 0;
+    m_nodesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(const AddEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = undoAction.GetEdge();
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(const ResetNodeAction& undoAction)
+{
+    m_nodes[undoAction.NodeId()] = undoAction.UpdatedNode();
+    m_nodesRTreeRequiresUpdate = true;
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(const ResetEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = undoAction.UpdatedEdge();
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(const DeleteEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = {constants::missing::uintValue, constants::missing::uintValue};
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(const DeleteNodeAction& undoAction)
+{
+    // undoAction.CommitEdges();
+    m_nodes[undoAction.NodeId()] = {constants::missing::doubleValue, constants::missing::doubleValue};
+    m_nodesRTreeRequiresUpdate = true;
+}
+
+void Mesh::CommitAction(NodeTranslationAction& undoAction)
+{
+    undoAction.Swap(m_nodes);
+}
+
+void Mesh::CommitAction(MeshConversionAction& undoAction)
+{
+    undoAction.Swap(m_nodes, m_projection);
+}
+
+void Mesh::RestoreAction(const AddNodeAction& undoAction)
+{
+    m_nodes[undoAction.NodeId()] = Point(constants::missing::doubleValue, constants::missing::doubleValue);
+    // TODO is this necessary, will it be set in Administrate?
+    // Need to collect the correct value here for this.
+    m_nodesNumEdges[undoAction.NodeId()] = 0;
+    m_nodesRTreeRequiresUpdate = true;
+}
+
+void Mesh::RestoreAction(const AddEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = {constants::missing::uintValue, constants::missing::uintValue};
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::RestoreAction(const ResetNodeAction& undoAction)
+{
+    m_nodes[undoAction.NodeId()] = undoAction.InitialNode();
+    m_nodesRTreeRequiresUpdate = true;
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::RestoreAction(const ResetEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = undoAction.InitialEdge();
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::RestoreAction(const DeleteEdgeAction& undoAction)
+{
+    m_edges[undoAction.EdgeId()] = undoAction.GetEdge();
+    m_edgesRTreeRequiresUpdate = true;
+}
+
+void Mesh::RestoreAction(const DeleteNodeAction& undoAction)
+{
+    m_nodes[undoAction.NodeId()] = undoAction.Node();
+    // TODO DO we need to assign the m_nodesNumEdges the length of the deleted edges array?
```

Comment:
Do you want to address this? If yes, please do. Otherwise remove the comment and file an issue.